File: python-packages/core/src/omigo_core/file_io_wrapper.py
"""FileReader / FileWriter class"""

import gzip
import zipfile

from omigo_core import s3_wrapper
from omigo_core import utils

# TODO: this is very inefficient. Use s3fs or write something better.
class FileWriter:
    """FileWriter class to write data into files"""
    def __init__(self, output_file_name, s3_region, aws_profile):
        self.output_file_name = output_file_name
        self.data = []
        self.s3_region = s3_region
        self.aws_profile = aws_profile

    def write(self, line):
        self.data.append(line)

    def close(self):
        output_zipf = None
        output_file = None
        if (self.output_file_name.startswith("s3://")):
            content = "".join(self.data)
            bucket_name, object_key = utils.split_s3_path(self.output_file_name)
            s3_wrapper.put_s3_file_with_text_content(bucket_name, object_key, content, self.s3_region, self.aws_profile)
        else:
            # construct output file
            if (self.output_file_name.endswith(".gz")):
                output_file = gzip.open(self.output_file_name, "wt")
                # write all the content
                for line in self.data:
                    output_file.write(line)
            elif (self.output_file_name.endswith(".zip")):
                output_zipf = zipfile.ZipFile(self.output_file_name, "w")
                output_file = output_zipf.open(self.output_file_name.split("/")[-1][0:-4], "w")
                # write all the content
                for line in self.data:
                    output_file.write(str.encode(line))
            else:
                output_file = open(self.output_file_name, "w")
                # write all the content
                for line in self.data:
                    output_file.write(line)

            # close
            if (output_file is not None):
                output_file.close()
            if (output_zipf is not None):
                output_zipf.close()

            # set data to None
            self.data = None

class TSVFileWriter:
    """FileWriter class to write data into files"""
    def __init__(self, s3_region, aws_profile):
        self.s3_region = s3_region
        self.aws_profile = aws_profile

    def save(self, xtsv, output_file_name):
        output_zipf = None
        output_file = None
        if (output_file_name.startswith("s3://")):
            content = xtsv.get_header() + "\n" + "\n".join(xtsv.get_data()) if (xtsv.num_rows() > 0) else xtsv.get_header()
            bucket_name, object_key = utils.split_s3_path(output_file_name)
            s3_wrapper.put_s3_file_with_text_content(bucket_name, object_key, content, self.s3_region, self.aws_profile)
        else:
            # construct output file
            if (output_file_name.endswith(".gz")):
                output_file = gzip.open(output_file_name, "wt")
                # write header
                output_file.write(xtsv.get_header() + "\n")
                # write all the content
                for line in xtsv.get_data():
                    output_file.write(line + "\n")
            elif (output_file_name.endswith(".zip")):
                output_zipf = zipfile.ZipFile(output_file_name, "w")
                output_file = output_zipf.open(output_file_name.split("/")[-1][0:-4], "w")
                # write header
                output_file.write(xtsv.get_header() + "\n")
                # write all the content. TODO: what is str.encode doing here?
                utils.warn_once("str.encode used. Not sure why") 
                for line in xtsv.get_data():
                    output_file.write(str.encode(line + "\n"))
            else:
                output_file = open(output_file_name, "w")
                # write header
                output_file.write(xtsv.get_header() + "\n")
                # write all the content
                for line in xtsv.get_data():
                    output_file.write(line + "\n")

            # close
            if (output_file is not None):
                output_file.close()
            if (output_zipf is not None):
                output_zipf.close()

class FileReader:
    """FileReader class to read data files"""
    def __init__(self, input_file_name, s3_region, aws_profile):
        pass

    def read(self):
        pass

    def close(self):
        pass

File: projects/lokki/java-scala/src/main/scala/omigo_core/FileIOWrapper.scala
package omigo_core
// import scala.jdk.CollectionConverters
import collection.JavaConverters._
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.File
import java.io.FileOutputStream
import java.nio.ByteBuffer
import java.util.Random
import java.util.zip.GZIPInputStream
import java.util.zip.GZIPOutputStream

class FileWriter(val output_file_name: String, val s3_region: String = null, val aws_profile: String = null) {
  // TODO: Not needed probably
}

class TSVFileWriter(val s3_region: String, val aws_profile: String) {
  def save(xtsv: TSV, output_file_name: String): Unit = {
    val content = List(xtsv.get_header(), xtsv.get_data().mkString("\n")).mkString("\n")

    // check s3 or local
    if (output_file_name.startsWith("s3://")) {
      val (bucket_name, object_key) = Utils.split_s3_path(output_file_name)
      S3Wrapper.put_s3_file_with_text_content(bucket_name, object_key, content, s3_region = s3_region, aws_profile = aws_profile)
      println("file saved to: " + output_file_name)
    } else {
      // get bytes
      var barr = content.getBytes()

      // check file extensions
      if (output_file_name.endsWith(".gz")) {
        val byteStream = new ByteArrayOutputStream(barr.length)
        val zipStream = new GZIPOutputStream(byteStream)
        zipStream.write(barr)
        zipStream.close()
        barr = byteStream.toByteArray()
      } else if (output_file_name.endsWith(".zip")) {
        throw new Exception("Not supported")
      }

      // open FileOutputStream and write
      val fos = new FileOutputStream(new File(output_file_name))
      fos.write(barr)
      fos.close()
      println("file saved to: " + output_file_name)
    }
  }
}

class FileReader {
  // TODO: Not needed
}

object TSVFileWriter {
  def main(args: Array[String]): Unit = {
    val xtsv = new TSV("col1", List("value1", "value2")) 
    val writer = new TSVFileWriter(null, null)
    writer.save(xtsv, args(0))
  }
}
